### PR TITLE
[stdlib] Fix `Int.MIN` error

### PR DESCRIPTION
**Pull request by helehex**
Fixes the compilation error you get when using `Int.MIN`

### CODE REVIEW

**Review by JoeLoser — COMMENTED** [review]
Can you share more about your system where this fails to compile? Is it a 32-bit system?

**Comment by helehex** [comment]
Host Information
  ================

  Target Triple: x86_64-unknown-linux
  CPU: skylake
  CPU Features: adx, aes, avx, avx2, bmi, bmi2, clflushopt, cmov, crc32, cx16, cx8, f16c, fma, fsgsbase, fxsr, invpcid, lzcnt, mmx, movbe, pclmul, popcnt, prfchw, rdrnd, rdseed, sahf, sgx, sse, sse2, sse3, sse4.1, sse4.2, ssse3, x87, xsave, xsavec, xsaveopt, xsaves

mojo 2024.4.818 (34399306)

modular 0.7.0 (487c2f88)

**Comment by helehex** [comment]
i was assuming this would happen regardless of platform because of the logic, maybe i was mistaken?

**Review by abduld — COMMENTED** [review]
can you add a test that Scalar[DType.index].MIN has a value. Also, would be great to consolidate these into an _is_index_32bit helper function to avoid duplication

**Comment by helehex** [comment]
sure no problem

**Comment by helehex** [comment]
how's that? it probably would've been fine to just `return sizeof[DType.index]() == sizeof[DType.int32]()`,
but figured i'd just be safe and explicit

**Comment by helehex** [comment]
oh there's one whitespace it doesn't like hang on

**Review by JoeLoser — APPROVED** [review]
Looks good, thank you for the contribution! 🎉 

Do you mind rebasing to resolve the conflict?  Happy to merge after that is taken care of.

File: stdlib/src/builtin/simd.mojo [hunk 1]
```diff
@@ -105,6 +105,29 @@ fn _unchecked_zero[type: DType, size: Int]() -> SIMD[type, size]:
     }
 
 
+@always_inline
+fn _is_index_32bit() -> Bool:
+    @parameter
+    if sizeof[DType.index]() == sizeof[DType.int32]():
+        return True
+    else:
+        return False
```
Comment (soraros) [inline stdlib/src/builtin/simd.mojo:114]: I don't think the param if is necessary, as the function itself is only called in a param if so it's guaranteed to be compile time.

Comment (helehex) [reply stdlib/src/builtin/simd.mojo:114]: it should probably go somewhere else aswell, like `dtype.mojo` or `info.mojo`

**Comment by helehex** [comment]
seems like someone already solved it and thats what the conflict is

**Comment by helehex** [comment]
maybe i'll move that helper function to `info.mojo`, then add a corresponding one in `dtype.mojo` that does both checks. does that sound good?

**Comment by helehex** [comment]
oh i just noticed the import for sys sizeof in `dtype.mojo` isn't being used

File: stdlib/src/sys/info.mojo
Comment (helehex) [inline stdlib/src/sys/info.mojo:1]: how about this?

File: stdlib/src/sys/info.mojo [hunk 2]
```diff
@@ -446,6 +446,32 @@ fn is_big_endian[
     ]
 
 
+@always_inline("nodebug")
```
Comment (helehex) [inline stdlib/src/sys/info.mojo:449]: how about this?

File: stdlib/src/sys/info.mojo [hunk 3]
```diff
@@ -446,6 +446,32 @@ fn is_big_endian[
...
+
+    Parameters:
+        target: The target architecture.
+
+    Returns:
+        True if the maximum integral value is 64 bit, False otherwise.
+    """
+    return sizeof[DType.index, target]() == sizeof[DType.int64, target]()
+
+
```
Comment (helehex) [inline stdlib/src/sys/info.mojo:474]: how about this?

File: stdlib/src/sys/info.mojo [hunk 4]
```diff
@@ -446,6 +446,32 @@ fn is_big_endian[
...
+fn is_32bit[target: __mlir_type.`!kgen.target` = _current_target()]() -> Bool:
+    """Returns True if the maximum integral value is 32 bit.
+
+    Parameters:
+        target: The target architecture.
+
+    Returns:
+        True if the maximum integral value is 32 bit, False otherwise.
+    """
+    return sizeof[DType.index, target]() == sizeof[DType.int32, target]()
```
Comment (helehex) [inline stdlib/src/sys/info.mojo:459]: what do you think about this?

Comment (JoeLoser) [reply stdlib/src/sys/info.mojo:459]: LGTM.  Seems like a much more natural place to put this function. 👍 

Thanks!

File: stdlib/src/builtin/simd.mojo [hunk 5]
```diff
@@ -2538,11 +2538,11 @@ fn _min_finite[type: DType]() -> Scalar[type]:
         return -128
     elif type == DType.int16:
         return -32768
-    elif type == DType.int32 or (type == DType.index and _is_32_bit_system()):
+    elif type == DType.int32 or type.is_index32():
         return -2147483648
-    elif type == DType.int64 or (
-        type == DType.index and not _is_32_bit_system()
-    ):
+    elif type == DType.float32:
```
Comment (JoeLoser) [inline stdlib/src/builtin/simd.mojo:2543]: **Suggestion** I think this rebase part is not correct.  We want to just fall through to the bottom part of `elif type.is_floating_point()` which we changed internally recently.

**Review by JoeLoser — APPROVED** [review]
Looks good still, but please fixup the one part about the rebase went slightly awry (correct, but not quite what we want).

File: stdlib/src/builtin/simd.mojo (hunk 5 above)
Comment (helehex) [reply stdlib/src/builtin/simd.mojo:2543]: oh i see

**Comment by helehex** [comment]
oh i didn't do a `rebase upstream/nightly` that time, should that matter? there's no conflicts

**Comment by JoeLoser** [comment]
> oh i didn't do a `rebase upstream/nightly` that time, should that matter? there's no conflicts

Nah, force pushing is totally fine.  No biggie.  I'll merge this now! 😄

**Pull request merged**